messaging.wit.md

```diff
@@ -0,0 +1,105 @@
+# `wasi-messaging` API
+
+## Interfaces
+
+```wit
+/// An interface grouping all necessary types for messaging.
+interface "wasi:messaging/types" {
+    /// An event type that follows the CloudEvents specification (https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md). We
+    /// assume the type of the data is a byte sequence. It is up to the data schema to determine what type of the data payload the event 
+    /// contains.
+    record event {
+    	specversion: string,
+    	ty: string,
+    	source: string,
+    	id: string,
+    	data: option<list<u8>>,
+    	datacontenttype: option<string>,
+    	dataschema: option<string>,
+    	subject: option<string>,
+    	time: option<string>,
+    }
+    
+    /// Channels specify where a published message should land. There are two types of channels:
+    /// - queue: competitive consumers, and
+    /// - topic: non-competitive consumers.
+    variant channel {
+    	queue(string),
+    	topic(string)
+    }
+    
+    /// An error type
+    type error = string
+}
+
+/// An interface for a producer.
+interface "wasi:messaging/pub" { 
+	/// Publishes an event to a channel.
+	publish: func(c: channel, e: event) -> result<_, error>
+}
+
+/// An interface for a generic consumer.
+interface "wasi:messaging/sub" {
+	/// Subscribes to a channel.
+  	subscribe: func(c: channel) -> result<_, error>
+
+	/// Unsubscribes from a channel.
+  	unsubscribe: func(c: channel) -> result<_, error>
+}
+
+/// An interface for a consumer relying on push-based message delivery.
+interface "wasi:messaging/push" {
+	/// Creates an on-receive handler push-based message delivery.
+  	on-receive: func(e: event) -> result<_, error>
+}
+
+/// An interface for a consumer relying on basic pull-based message delivery.
+interface "wasi:messaging/basic/pull" {
+	/// Pulls a message.
```

Comment:
in wasmCloud, a majority use case is `receive_timeout()`, which returns `result<event,Error>` where Error can be a timeout error or some other error. In this interface, the timeout option is probably applicable to both pull and push consumers.

Reply:
That's right~ I actually meant to include a `time-to-live` parameter for `receive` but it ended up escaping me. That said, I'm not sure that is applicable for push-based message delivery as there's no waiting from the user side.